# Designing and Creating a Database

### Introduction

In this project, we will use data from Major League Baseball games from [Retrosheet](http://www.retrosheet.org/). The main file is game_log.csv, an file that was compiled by Dataquest from 127 CSV files from retrosheet. This file contains hundreds of data points on each game.

We will also use three 'helper' files, also sourced from Retrosheet: park_codes.csv, person_codes.csv, team_codes.csv.

The learning objectives of this project are to design, create and populate a normalized SQL database using Python and SQLite.

### Understanding the data

In [25]:
import sqlite3
import pandas as pd
import csv

pd.set_option('max_columns', 180)
pd.set_option('max_rows', 200000)
pd.set_option('max_colwidth', 5000)

In [26]:
log = pd.read_csv("game_log.csv", low_memory=False)
log.head()

date  number_of_game day_of_week v_name v_league  v_game_number h_name  \
0  18710504               0         Thu    CL1      NaN              1    FW1   
1  18710505               0         Fri    BS1      NaN              1    WS3   
2  18710506               0         Sat    CL1      NaN              2    RC1   
3  18710508               0         Mon    CL1      NaN              3    CH1   
4  18710509               0         Tue    BS1      NaN              2    TRO   

  h_league  h_game_number  v_score  h_score  length_outs day_night completion  \
0      NaN              1        0        2         54.0         D        NaN   
1      NaN              1       20       18         54.0         D        NaN   
2      NaN              1       12        4         54.0         D        NaN   
3      NaN              1       12       14         54.0         D        NaN   
4      NaN              1        9        5         54.0         D        NaN   

  forefeit protest park_id  attendance  length_minutes v_line_score  \
0      NaN     NaN   FOR01       200.0           120.0    000000000   
1      NaN     NaN   WAS01      5000.0           145.0    107000435   
2      NaN     NaN   RCK01      1000.0           140.0    610020003   
3      NaN     NaN   CHI01      5000.0           150.0    101403111   
4      NaN     NaN   TRO01      3250.0           145.0    000002232   

  h_line_score  v_at_bats  v_hits  v_doubles  v_triples  v_homeruns  v_rbi  \
0    010010000       30.0     4.0        1.0        0.0         0.0    0.0   
1    640113030       41.0    13.0        1.0        2.0         0.0   13.0   
2    010020100       49.0    11.0        1.0        1.0         0.0    8.0   
3    077000000       46.0    15.0        2.0        1.0         2.0   10.0   
4    101003000       46.0    17.0        4.0        1.0         0.0    6.0   

   v_sacrifice_hits  v_sacrifice_flies  v_hit_by_pitch  v_walks  \
0               0.0                0.0             0.0      1.0   
1               0.0                0.0             0.0     18.0   
2               0.0                0.0             0.0      0.0   
3               0.0                0.0             0.0      0.0   
4               0.0                0.0             0.0      2.0   

   v_intentional_walks  v_strikeouts  v_stolen_bases  v_caught_stealing  \
0                  NaN           6.0             1.0                NaN   
1                  NaN           5.0             3.0                NaN   
2                  NaN           1.0             0.0                NaN   
3                  NaN           1.0             0.0                NaN   
4                  NaN           0.0             1.0                NaN   

   v_grounded_into_double  v_first_catcher_interference  v_left_on_base  \
0                    -1.0                           NaN             4.0   
1                    -1.0                           NaN            12.0   
2                    -1.0                           NaN            10.0   
3                    -1.0                           NaN             7.0   
4                    -1.0                           NaN            12.0   

   v_pitchers_used  v_individual_earned_runs  v_team_earned_runs  \
0              1.0                       1.0                 1.0   
1              1.0                       6.0                 6.0   
2              1.0                       0.0                 0.0   
3              1.0                       6.0                 6.0   
4              1.0                       2.0                 2.0   

   v_wild_pitches  v_balks  v_putouts  v_assists  v_errors  v_passed_balls  \
0             0.0      0.0       27.0        9.0       0.0             3.0   
1             1.0      0.0       27.0       13.0      10.0             1.0   
2             2.0      0.0       27.0       12.0       8.0             5.0   
3             0.0      0.0       27.0       15.0      11.0             6.0   
4             0.0      0.0       

In [27]:
log.shape

(171907, 161)

The game log has a record of over 170,000 games. It looks like these games are chronologically ordered and occur between 1871 and 2016.

For each game we have:

- general information on the game
- team level stats for each team
- a list of players from each team, numbered, with their defensive positions
- the umpires that officiated the game
- some 'awards', like winning and losing pitcher

It's worth noting that there is no natural primary key column for this table.

In [28]:
park = pd.read_csv("park_codes.csv")
park.head()

park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   

                                        aka       city state       start  \
0                                       NaN     Albany    NY  09/11/1880   
1                                       NaN    Altoona    PA  04/30/1884   
2             Edison Field; Anaheim Stadium    Anaheim    CA  04/19/1966   
3                                       NaN  Arlington    TX  04/21/1972   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX  04/11/1994   

          end league                                           notes  
0  05/30/1882     NL  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1  05/31/1884     UA                                             NaN  
2         NaN     AL                                             NaN  
3  10/03/1993     AL                                             NaN  
4         NaN     AL                                             NaN

In [29]:
park.shape

(252, 9)

This seems to be a list of all baseball parks (252 in total). There are IDs which seem to match with the game log, as well as names, nicknames, city and league.

In [30]:
person = pd.read_csv("person_codes.csv")
person.head()

id     last   first player_debut mgr_debut coach_debut ump_debut
0  aardd001  Aardsma   David   04/06/2004       NaN         NaN       NaN
1  aaroh101    Aaron    Hank   04/13/1954       NaN         NaN       NaN
2  aarot101    Aaron  Tommie   04/10/1962       NaN  04/06/1979       NaN
3  aased001     Aase     Don   07/26/1977       NaN         NaN       NaN
4  abada001     Abad    Andy   09/10/2001       NaN         NaN       NaN

In [31]:
person.shape

(20494, 7)

This is a list of people with IDs. The IDs look like they match up with those used in the game log. There are debut dates, for players, managers, coaches and umpires. We can see that some people might have been one or more of these roles.

It also looks like coaches and managers are two different things in baseball. After some research, managers are what would be called a 'coach' or 'head coach' in other sports, and coaches are more specialized, like base coaches. It also seems like coaches aren't recorded in the game log.

In [32]:
team = pd.read_csv("team_codes.csv")
team.head()

team_id league  start   end       city         nickname franch_id  seq
0     ALT     UA   1884  1884    Altoona  Mountain Cities       ALT    1
1     ARI     NL   1998     0    Arizona     Diamondbacks       ARI    1
2     BFN     NL   1879  1885    Buffalo           Bisons       BFN    1
3     BFP     PL   1890  1890    Buffalo           Bisons       BFP    1
4     BL1    NaN   1872  1874  Baltimore         Canaries       BL1    1

In [33]:
team.shape

(150, 8)

This seems to be a list of all teams, with team_ids which seem to match the game log. There is a franch_id and it's not clear the difference between this franch_id and the team_id. To understand the difference, we will look further into the data:

In [34]:
team["franch_id"].value_counts().head()

BS1    4
TRN    3
LAA    3
BR3    3
SE1    3
Name: franch_id, dtype: int64

Some franch_id repeats itself. Let's look in more details into the first one (BS1):

In [35]:
team[team["franch_id"] == 'BS1']

team_id league  start   end       city nickname franch_id  seq
21     BS1    NaN   1871  1875     Boston   Braves       BS1    1
22     BSN     NL   1876  1952     Boston   Braves       BS1    2
23     MLN     NL   1953  1965  Milwaukee   Braves       BS1    3
24     ATL     NL   1966     0    Atlanta   Braves       BS1    4

It appears that teams move between leagues and cities. The team_id changes when this happens, franch_id (which is probably 'Franchise') helps us tie all of this together.

### Importing data into SQLite

First, we will define some functions that will be used in the next steps of the project.

In [36]:
DB = "mlb.db"

def run_query(q):
    with sqlite3.connect(DB) as conn:
        return pd.read_sql(q,conn)

def run_command(c):
    with sqlite3.connect(DB) as conn:
        conn.execute('PRAGMA foreign_keys = ON;')
        conn.isolation_level = None
        conn.execute(c)

def show_tables():
    q = '''
    SELECT
        name,
        type
    FROM sqlite_master
    WHERE type IN ("table","view");
    '''
    return run_query(q)

Now, we will create the tables:

In [37]:
tables  =  {
    "game_log": log,
    "person_codes": person,
    "team_codes": team,
    "park_codes": park
}

with sqlite3.connect(DB) as conn:    
    for name, data in tables.items():
        conn.execute("DROP TABLE IF EXISTS {};".format(name))
        data.to_sql(name,conn,index=False)

In [38]:
show_tables()

name   type
0  person_codes  table
1    park_codes  table
2      game_log  table
3    team_codes  table

Then, we will created a game_id field combining  8 characters for the year (4 characters), month (2) and day (2) of the game, the first three characters of the home team and then the last digit could be 0, 1 or 2, representing if this is a single game (0), the first game played on the day (1) or the second (2).

For example, the id "19830408ATL0" represents the single game played by the Atlanta Braves on  1983-04-08. 

In [40]:
c1 = """
ALTER TABLE game_log
ADD COLUMN game_id TEXT;
"""

# try/except loop since ALTER TABLE
# doesn't support IF NOT EXISTS
try:
    run_command(c1)
except:
    pass

c2 = """
UPDATE game_log
SET game_id = date || h_name || number_of_game
WHERE game_id IS NULL; 
"""

run_command(c2)

q = """
SELECT
    game_id,
    date,
    h_name,
    number_of_game
FROM game_log
LIMIT 5;
"""

run_query(q)

game_id      date h_name  number_of_game
0  18710504FW10  18710504    FW1               0
1  18710505WS30  18710505    WS3               0
2  18710506RC10  18710506    RC1               0
3  18710508CH10  18710508    CH1               0
4  18710509TRO0  18710509    TRO               0

### Normalization opportunities

We found some opportunities to normalize where there is repetition in columns, redundant data and where non-primary key columns are attributes of the primary key. 

The following are opportunities for normalization of our data:

- In person_codes, all the debut dates will be able to be reproduced using game log data.
- In team_codes, the start, end and sequence columns will be able to be reproduced using game log data.
- In park_codes, the start and end years will be able to be reproduced using game log data. While technically the state is an attribute of the city, we might not want to have a an incomplete city/state table so we will leave this in.
- There are lots of places in game log where we have a player ID followed by the players name. We will be able to remove this and use the name data in person_codes
- In game_log, all offensive and defensive stats are repeated for the home team and the visiting team. We could break these out and have a table that lists each game twice, one for each team, and cut out this column repetition.
- Similarly, in game_log, we have a listing for 9 players on each team with their positions - we can remove these and have one table that tracks player appearances and their positions.
- We can do a similar thing with the umpires from game_log, instead of listing all four positions as columns, we can put the umpires either in their own table or make one table for players, umpires and managers.
- We have several awards in game_log like winning pitcher and losing pitcher. We can either break these out into their own table, have a table for awards, or combine the awards in with general appearances like the players and umpires.

### Planning the database schema

This normalized database schema was created using [DBDesigner.net](https://dbdesigner.net/)

![Schema](https://github.com/dataquestio/solutions/raw/84a5fe20efcb4f14a8d508ff4b4c98d1ab80923a/images/schema-screenshot.png)

### Creating Tables Without Foreign Key Relations 

To create those tables, we will use three steps:

- Creating the table
- Populating the table with data coming from the original tables. In this step, we will take caution to change the columns order and to eventually process data using CASE WHEN.
- Selecting the first 5 rows of the table for display purposes

In [41]:
c1 = """
CREATE TABLE IF NOT EXISTS person (
    person_id TEXT PRIMARY KEY,
    first_name TEXT,
    last_name TEXT
);
"""

c2 = """
INSERT OR IGNORE INTO person
SELECT
    id,
    first,
    last
FROM person_codes;
"""

q = """
SELECT * FROM person
LIMIT 5;
"""

run_command(c1)
run_command(c2)
run_query(q)

person_id first_name last_name
0  aardd001      David   Aardsma
1  aaroh101       Hank     Aaron
2  aarot101     Tommie     Aaron
3  aased001        Don      Aase
4  abada001       Andy      Abad

In [45]:
c1 = """
CREATE TABLE IF NOT EXISTS park (
    park_id TEXT PRIMARY KEY,
    name TEXT,
    nickname TEXT,
    city TEXT,
    state TEXT,
    notes TEXT
);
"""

c2 = """
INSERT OR IGNORE INTO park
SELECT
    park_id,
    name,
    aka,
    city,
    state,
    notes
FROM park_codes;
"""

q = """
SELECT * FROM park
LIMIT 5;
"""

run_command(c1)
run_command(c2)
run_query(q)

park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   

                                   nickname       city state  \
0                                      None     Albany    NY   
1                                      None    Altoona    PA   
2             Edison Field; Anaheim Stadium    Anaheim    CA   
3                                      None  Arlington    TX   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX   

                                            notes  
0  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1                                            None  
2                                            None  
3                                            None  
4                                            None

In [46]:
c1 = """
CREATE TABLE IF NOT EXISTS league (
    league_id TEXT PRIMARY KEY,
    name TEXT
);
"""

c2 = """
INSERT OR IGNORE INTO league
VALUES
    ("NL", "National League"),
    ("AL", "American League"),
    ("AA", "American Association"),
    ("FL", "Federal League"),
    ("PL", "Players League"),
    ("UA", "Union Association")
;
"""

q = """
SELECT * FROM league
"""

run_command(c1)
run_command(c2)
run_query(q)

league_id                  name
0        NL       National League
1        AL       American League
2        AA  American Association
3        FL        Federal League
4        PL        Players League
5        UA     Union Association

In [47]:
c1 = "DROP TABLE IF EXISTS appearance_type;"

run_command(c1)

c2 = """
CREATE TABLE appearance_type (
    appearance_type_id TEXT PRIMARY KEY,
    name TEXT,
    category TEXT
);
"""
run_command(c2)

appearance_type = pd.read_csv('appearance_type.csv')

with sqlite3.connect('mlb.db') as conn:
    appearance_type.to_sql('appearance_type',
                           conn,
                           index=False,
                           if_exists='append')

q = """
SELECT * FROM appearance_type;
"""

run_query(q)

appearance_type_id                name category
0                  O1            Batter 1  offense
1                  O2            Batter 2  offense
2                  O3            Batter 3  offense
3                  O4            Batter 4  offense
4                  O5            Batter 5  offense
5                  O6            Batter 6  offense
6                  O7            Batter 7  offense
7                  O8            Batter 8  offense
8                  O9            Batter 9  offense
9                  D1             Pitcher  defense
10                 D2             Catcher  defense
11                 D3            1st Base  defense
12                 D4            2nd Base  defense
13                 D5            3rd Base  defense
14                 D6           Shortstop  defense
15                 D7          Left Field  defense
16                 D8        Center Field  defense
17                 D9         Right Field  defense
18                D10    Unknown Position  defense
19                UHP          Home Plate   umpire
20                U1B          First Base   umpire
21                U2B         Second Base   umpire
22                U3B          Third Base   umpire
23                ULF          Left Field   umpire
24                URF         Right Field   umpire
25                 MM             Manager  manager
26                AWP     Winning Pitcher    award
27                ALP      Losing Pitcher    award
28                ASP      Saving Pitcher    award
29                AWB  Winning RBI Batter    award
30                PSP    Starting Pitcher  pitcher

### Creating Team and Game Tables

Before creating the two appearances tables, we need to created the game and team tables.

In [48]:
c1 = """
CREATE TABLE IF NOT EXISTS team (
    team_id TEXT PRIMARY KEY,
    league_id TEXT,
    city TEXT,
    nickname TEXT,
    franch_id TEXT,
    FOREIGN KEY (league_id) REFERENCES league(league_id)
);
"""

c2 = """
INSERT OR IGNORE INTO team
SELECT
    team_id,
    league,
    city,
    nickname,
    franch_id
FROM team_codes;
"""

q = """
SELECT * FROM team
LIMIT 5;
"""

run_command(c1)
run_command(c2)
run_query(q)

team_id league_id       city         nickname franch_id
0     ALT        UA    Altoona  Mountain Cities       ALT
1     ARI        NL    Arizona     Diamondbacks       ARI
2     BFN        NL    Buffalo           Bisons       BFN
3     BFP        PL    Buffalo           Bisons       BFP
4     BL1      None  Baltimore         Canaries       BL1

In [49]:
c1 = """
CREATE TABLE IF NOT EXISTS game (
    game_id TEXT PRIMARY KEY,
    date TEXT,
    number_of_game INTEGER,
    park_id TEXT,
    length_outs INTEGER,
    day BOOLEAN,
    completion TEXT,
    forefeit TEXT,
    protest TEXT,
    attendance INTEGER,
    legnth_minutes INTEGER,
    additional_info TEXT,
    acquisition_info TEXT,
    FOREIGN KEY (park_id) REFERENCES park(park_id)
);
"""

c2 = """
INSERT OR IGNORE INTO game
SELECT
    game_id,
    date,
    number_of_game,
    park_id,
    length_outs,
    CASE
        WHEN day_night = "D" THEN 1
        WHEN day_night = "N" THEN 0
        ELSE NULL
        END
        AS day,
    completion,
    forefeit,
    protest,
    attendance,
    length_minutes,
    additional_info,
    acquisition_info
FROM game_log;
"""

q = """
SELECT * FROM game
LIMIT 5;
"""

run_command(c1)
run_command(c2)
run_query(q)

game_id      date  number_of_game park_id  length_outs  day  \
0  18710504FW10  18710504               0   FOR01           54    1   
1  18710505WS30  18710505               0   WAS01           54    1   
2  18710506RC10  18710506               0   RCK01           54    1   
3  18710508CH10  18710508               0   CHI01           54    1   
4  18710509TRO0  18710509               0   TRO01           54    1   

  completion forefeit protest  attendance  legnth_minutes additional_info  \
0       None     None    None         200             120            None   
1       None     None    None        5000             145            HTBF   
2       None     None    None        1000             140            None   
3       None     None    None        5000             150            None   
4       None     None    None        3250             145            HTBF   

  acquisition_info  
0                Y  
1                Y  
2                Y  
3                Y  
4                Y

### Creating the Team Appearance Table

In [50]:
c1 = """
CREATE TABLE IF NOT EXISTS team_appearance (
    team_id TEXT,
    game_id TEXT,
    home BOOLEAN,
    league_id TEXT,
    score INTEGER,
    line_score TEXT,
    at_bats INTEGER,
    hits INTEGER,
    doubles INTEGER,
    triples INTEGER,
    homeruns INTEGER,
    rbi INTEGER,
    sacrifice_hits INTEGER,
    sacrifice_flies INTEGER,
    hit_by_pitch INTEGER,
    walks INTEGER,
    intentional_walks INTEGER,
    strikeouts INTEGER,
    stolen_bases INTEGER,
    caught_stealing INTEGER,
    grounded_into_double INTEGER,
    first_catcher_interference INTEGER,
    left_on_base INTEGER,
    pitchers_used INTEGER,
    individual_earned_runs INTEGER,
    team_earned_runs INTEGER,
    wild_pitches INTEGER,
    balks INTEGER,
    putouts INTEGER,
    assists INTEGER,
    errors INTEGER,
    passed_balls INTEGER,
    double_plays INTEGER,
    triple_plays INTEGER,
    PRIMARY KEY (team_id, game_id),
    FOREIGN KEY (team_id) REFERENCES team(team_id),
    FOREIGN KEY (game_id) REFERENCES game(game_id),
    FOREIGN KEY (team_id) REFERENCES team(team_id)
);
"""

run_command(c1)

c2 = """
INSERT OR IGNORE INTO team_appearance
    SELECT
        h_name,
        game_id,
        1 AS home,
        h_league,
        h_score,
        h_line_score,
        h_at_bats,
        h_hits,
        h_doubles,
        h_triples,
        h_homeruns,
        h_rbi,
        h_sacrifice_hits,
        h_sacrifice_flies,
        h_hit_by_pitch,
        h_walks,
        h_intentional_walks,
        h_strikeouts,
        h_stolen_bases,
        h_caught_stealing,
        h_grounded_into_double,
        h_first_catcher_interference,
        h_left_on_base,
        h_pitchers_used,
        h_individual_earned_runs,
        h_team_earned_runs,
        h_wild_pitches,
        h_balks,
        h_putouts,
        h_assists,
        h_errors,
        h_passed_balls,
        h_double_plays,
        h_triple_plays
    FROM game_log

UNION

    SELECT    
        v_name,
        game_id,
        0 AS home,
        v_league,
        v_score,
        v_line_score,
        v_at_bats,
        v_hits,
        v_doubles,
        v_triples,
        v_homeruns,
        v_rbi,
        v_sacrifice_hits,
        v_sacrifice_flies,
        v_hit_by_pitch,
        v_walks,
        v_intentional_walks,
        v_strikeouts,
        v_stolen_bases,
        v_caught_stealing,
        v_grounded_into_double,
        v_first_catcher_interference,
        v_left_on_base,
        v_pitchers_used,
        v_individual_earned_runs,
        v_team_earned_runs,
        v_wild_pitches,
        v_balks,
        v_putouts,
        v_assists,
        v_errors,
        v_passed_balls,
        v_double_plays,
        v_triple_plays
    from game_log;
"""

run_command(c2)

q = """
SELECT * FROM team_appearance
WHERE game_id = (
                 SELECT MIN(game_id) from game
                )
   OR game_id = (
                 SELECT MAX(game_id) from game
                )
ORDER By game_id, home;
"""

run_query(q)

team_id       game_id  home league_id  score line_score  at_bats  hits  \
0     CL1  18710504FW10     0      None      0  000000000       30     4   
1     FW1  18710504FW10     1      None      2  010010000       31     4   
2     MIA  20161002WAS0     0        NL      7  000230020       38    14   
3     WAS  20161002WAS0     1        NL     10  03023002x       30    10   

   doubles  triples  homeruns  rbi  sacrifice_hits  sacrifice_flies  \
0        1        0         0    0               0                0   
1        1        0         0    2               0                0   
2        1        1         2    7               1                0   
3        2        0         1   10               1                1   

   hit_by_pitch  walks  intentional_walks  strikeouts  stolen_bases  \
0             0      1                NaN           6             1   
1             0      1                NaN           0             0   
2             0      3                2.0          10             1   
3             1      8                0.0           3             2   

   caught_stealing  grounded_into_double  first_catcher_interference  \
0              NaN                    -1                         NaN   
1              NaN                    -1                         NaN   
2              1.0                     1                         0.0   
3              0.0                     1                         0.0   

   left_on_base  pitchers_used  individual_earned_runs  team_earned_runs  \
0             4              1                       1                 1   
1             3              1                       0                 0   
2             8              7                      10                10   
3             7              6                       7                 7   

   wild_pitches  balks  putouts  assists  errors  passed_balls  double_plays  \
0             0      0       27        9       0             3             0   
1             0      0       27        3       3             1             1   
2             1      0       24       11       0             0             1   
3             1      0       27       11       0             0             1   

   triple_plays  
0             0  
1             0  
2             0  
3             0

### Creating the Person Appearance Table

In [51]:
c0 = "DROP TABLE IF EXISTS person_appearance"

run_command(c0)

c1 = """
CREATE TABLE person_appearance (
    appearance_id INTEGER PRIMARY KEY,
    person_id TEXT,
    team_id TEXT,
    game_id TEXT,
    appearance_type_id,
    FOREIGN KEY (person_id) REFERENCES person(person_id),
    FOREIGN KEY (team_id) REFERENCES team(team_id),
    FOREIGN KEY (game_id) REFERENCES game(game_id),
    FOREIGN KEY (appearance_type_id) REFERENCES appearance_type(appearance_type_id)
);
"""

c2 = """
INSERT OR IGNORE INTO person_appearance (
    game_id,
    team_id,
    person_id,
    appearance_type_id
) 
    SELECT
        game_id,
        NULL,
        hp_umpire_id,
        "UHP"
    FROM game_log
    WHERE hp_umpire_id IS NOT NULL    

UNION

    SELECT
        game_id,
        NULL,
        [1b_umpire_id],
        "U1B"
    FROM game_log
    WHERE "1b_umpire_id" IS NOT NULL

UNION

    SELECT
        game_id,
        NULL,
        [2b_umpire_id],
        "U2B"
    FROM game_log
    WHERE [2b_umpire_id] IS NOT NULL

UNION

    SELECT
        game_id,
        NULL,
        [3b_umpire_id],
        "U3B"
    FROM game_log
    WHERE [3b_umpire_id] IS NOT NULL

UNION

    SELECT
        game_id,
        NULL,
        lf_umpire_id,
        "ULF"
    FROM game_log
    WHERE lf_umpire_id IS NOT NULL

UNION

    SELECT
        game_id,
        NULL,
        rf_umpire_id,
        "URF"
    FROM game_log
    WHERE rf_umpire_id IS NOT NULL

UNION

    SELECT
        game_id,
        v_name,
        v_manager_id,
        "MM"
    FROM game_log
    WHERE v_manager_id IS NOT NULL

UNION

    SELECT
        game_id,
        h_name,
        h_manager_id,
        "MM"
    FROM game_log
    WHERE h_manager_id IS NOT NULL

UNION

    SELECT
        game_id,
        CASE
            WHEN h_score > v_score THEN h_name
            ELSE v_name
            END,
        winning_pitcher_id,
        "AWP"
    FROM game_log
    WHERE winning_pitcher_id IS NOT NULL

UNION

    SELECT
        game_id,
        CASE
            WHEN h_score < v_score THEN h_name
            ELSE v_name
            END,
        losing_pitcher_id,
        "ALP"
    FROM game_log
    WHERE losing_pitcher_id IS NOT NULL

UNION

    SELECT
        game_id,
        CASE
            WHEN h_score > v_score THEN h_name
            ELSE v_name
            END,
        saving_pitcher_id,
        "ASP"
    FROM game_log
    WHERE saving_pitcher_id IS NOT NULL

UNION

    SELECT
        game_id,
        CASE
            WHEN h_score > v_score THEN h_name
            ELSE v_name
            END,
        winning_rbi_batter_id,
        "AWB"
    FROM game_log
    WHERE winning_rbi_batter_id IS NOT NULL

UNION

    SELECT
        game_id,
        v_name,
        v_starting_pitcher_id,
        "PSP"
    FROM game_log
    WHERE v_starting_pitcher_id IS NOT NULL

UNION

    SELECT
        game_id,
        h_name,
        h_starting_pitcher_id,
        "PSP"
    FROM game_log
    WHERE h_starting_pitcher_id IS NOT NULL;
"""

template = """
INSERT INTO person_appearance (
    game_id,
    team_id,
    person_id,
    appearance_type_id
) 
    SELECT
        game_id,
        {hv}_name,
        {hv}_player_{num}_id,
        "O{num}"
    FROM game_log
    WHERE {hv}_player_{num}_id IS NOT NULL

UNION

    SELECT
        game_id,
        {hv}_name,
        {hv}_player_{num}_id,
        "D" || CAST({hv}_player_{num}_def_pos AS INT)
    FROM game_log
    WHERE {hv}_player_{num}_id IS NOT NULL;
"""

run_command(c1)
run_command(c2)

for hv in ["h","v"]:
    for num in range(1,10):
        query_vars = {
            "hv": hv,
            "num": num
        }
        run_command(template.format(**query_vars))

In [52]:
q = """
SELECT
    pa.*,
    at.name,
    at.category
FROM person_appearance pa
INNER JOIN appearance_type at on at.appearance_type_id = pa.appearance_type_id
WHERE PA.game_id = (
                   SELECT max(game_id)
                    FROM person_appearance
                   )
ORDER BY team_id, appearance_type_id
"""

run_query(q)

appearance_id person_id team_id       game_id appearance_type_id  \
0         1646109  porta901    None  20161002WAS0                U1B   
1         1646108  onorb901    None  20161002WAS0                U2B   
2         1646107  kellj901    None  20161002WAS0                U3B   
3         1646110  tumpj901    None  20161002WAS0                UHP   
4         1646111  brica001     MIA  20161002WAS0                ALP   
5         6716279  koeht001     MIA  20161002WAS0                 D1   
6         4744553  telit001     MIA  20161002WAS0                 D2   
7         5589581  bourj002     MIA  20161002WAS0                 D3   
8         4462877  gordd002     MIA  20161002WAS0                 D4   
9         5026229  pradm001     MIA  20161002WAS0                 D5   
10        6434609  hecha001     MIA  20161002WAS0                 D6   
11        5871257  scrux001     MIA  20161002WAS0                 D7   
12        5307905  yelic001     MIA  20161002WAS0                 D8   
13        6152933  hoodd001     MIA  20161002WAS0                 D9   
14        1646113  mattd001     MIA  20161002WAS0                 MM   
15        4462878  gordd002     MIA  20161002WAS0                 O1   
16        4744554  telit001     MIA  20161002WAS0                 O2   
17        5026230  pradm001     MIA  20161002WAS0                 O3   
18        5307906  yelic001     MIA  20161002WAS0                 O4   
19        5589582  bourj002     MIA  20161002WAS0                 O5   
20        5871258  scrux001     MIA  20161002WAS0                 O6   
21        6152934  hoodd001     MIA  20161002WAS0                 O7   
22        6434610  hecha001     MIA  20161002WAS0                 O8   
23        6716280  koeht001     MIA  20161002WAS0                 O9   
24        1646112  koeht001     MIA  20161002WAS0                PSP   
25        1646116  melam001     WAS  20161002WAS0                ASP   
26        1646115  difow001     WAS  20161002WAS0                AWB   
27        1646117  schem001     WAS  20161002WAS0                AWP   
28        4181201  schem001     WAS  20161002WAS0                 D1   
29        3899525  lobaj001     WAS  20161002WAS0                 D2   
30        2772821  zimmr001     WAS  20161002WAS0                 D3   
31        3336173  difow001     WAS  20161002WAS0                 D4   
32        3054497  drews001     WAS  20161002WAS0                 D5   
33        3617849  espid001     WAS  20161002WAS0                 D6   
34        2209469  reveb001     WAS  20161002WAS0                 D7   
35        1927793  turnt001     WAS  20161002WAS0                 D8   
36        2491145  harpb003     WAS  20161002WAS0                 D9   
37        1646114  baked002     WAS  20161002WAS0                 MM   
38        1927794  turnt001     WAS  20161002WAS0                 O1   
39        2209470  reveb001     WAS  20161002WAS0                 O2   
40        2491146  harpb003     WAS  20161002WAS0                 O3   
41        2772822  zimmr001     WAS  20161002WAS0                 O4   
42        3054498  drews001     WAS  20161002WAS0                 O5   
43        3336174  difow001     WAS  20161002WAS0                 O6   
44        3617850  espid001     WAS  20161002WAS0                 O7   
45        3899526  lobaj001     WAS  20161002WAS0                 O8   
46        4181202  schem001     WAS  20161002WAS0                 O9   
47        1646118  schem001     WAS  20161002WAS0                PSP   

                  name category  
0           First Base   umpire  
1          Second Base   umpire  
2           Third Base   umpire  
3           Home Plate   umpire  
4       Losing Pitcher    award  
5              Pitcher  defense  
6              Catcher  defense  
7             1st Base  defense  
8             2nd Base  defense  
9             3rd Base  defense  
10           Shortstop  defense  
11          Left Field  defense  
12        Center Field  defense  

### Removing the Original Tables

After we created all the tables, we can drop the tables used to hold our unnormalized data:

- game_log
- park_codes
- team_codes
- person_codes

In [53]:
tables = [
    "game_log",
    "park_codes",
    "team_codes",
    "person_codes"
]

for t in tables:
    c = '''
    DROP TABLE {}
    '''.format(t)
    
    run_command(c)

show_tables()

name   type
0             person  table
1               game  table
2               park  table
3             league  table
4    appearance_type  table
5               team  table
6    team_appearance  table
7  person_appearance  table